File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
class Solution {
    public int[] prevSmaller(int[] heights ){
        int n=heights.length;
        int[] res=new int[n];
        Stack<Integer>st=new Stack<>();
        st.push(0);
        res[0]=-1;
        for(int i=1;i<n;i++){
    if(heights[st.peek()]<heights[i]){
                res[i]=st.peek();
                 st.push(i);
            }
            else{
                while(!st.isEmpty() &&(heights[st.peek()]>=heights[i])){st.pop();
            }
                if(st.isEmpty()){
                    res[i]=-1;
                }
                else{
                    res[i]=st.peek();
                }
        
        st.push(i);}
        }
            return res;
        }
        public int[] nextSmaller(int[] heights){
            int n=heights.length;
        int[] res=new int[n];
        Stack<Integer>st=new Stack<>();
        st.push(n-1);
        res[n-1]=n;
        for(int i=n-2;i>=0;i--){
    if(heights[st.peek()]<heights[i]){
                res[i]=st.peek();
                 st.push(i);
            }
            else{
                while(!st.isEmpty() && heights[st.peek()]>=heights[i]){st.pop();
            }
                if(st.isEmpty()){
                    res[i]=n;
                }
                else{
                    res[i]=st.peek();
                }
                st.push(i);
        }
        }
            return res;
    
        }
    public int largestRectangleArea(int[] heights) {
        int n=heights.length;
        int[] prevS=prevSmaller(heights);
        int[] nextS= nextSmaller(heights);
        int maxArea=0;
        for(int i=0;i<n;i++){
            int width=nextS[i]-prevS[i]-1;
            int area= width*heights[i];
            maxArea=Math.max(maxArea,area);
        }
        return maxArea;
    }
}

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    vector<int> prevSmaller(vector<int>heights){
     int n=heights.size();
     stack<int>s;
     vector<int>res(n,-1);
     //default -1;
     res[0]=-1;
     s.push(0);
     for(int i=1;i<n;i++){
        if(heights[i]<=heights[s.top()]){
            while(!s.empty() && heights[i]<=heights[s.top()]){
                s.pop();
            }
            if(!s.empty()){
                res[i]=s.top();
            }
            else{
                res[i]=-1;
            }
            s.push(i);
        }
        else{
            res[i]=s.top();
            s.push(i);
        }
     }
     return res;
    }

     vector<int> nextSmaller(vector<int>heights){
        int n=heights.size();
        stack<int>st;
        vector<int>res(n,n);
        res[n-1]=n;
        //defaut value is n
        st.push(n-1);
        for(int i=n-2;i>=0;i--){
            if(heights[i]>heights[st.top()]){
                res[i]=st.top();
                st.push(i);
            }
            else{
                while(!st.empty() && heights[i]<=heights[st.top()]){
                    st.pop();
                }
                if(st.empty()){
                    res[i]=n;
                }
                else{
                    res[i]=st.top();
                }
                st.push(i);
            }
        }
        return res;
    }

    int largestRectangleArea(vector<int>& heights) {
        vector<int>prevS = prevSmaller(heights);
        vector<int>nextS= nextSmaller(heights);
      int n=heights.size();
        int maxArea=0;
        for(int i=0;i<n;i++){
            int wid = nextS[i]-prevS[i]-1;
            int area=wid*heights[i];
            maxArea=max(maxArea,area);
        }
        return maxArea;
    }
};